lib/Custom/material_bottom_navigation_scaffold.dart

```diff
@@ -0,0 +1,190 @@
+    }
+  }
+
+  @override
+  void dispose() {
+    _animationControllers.forEach(
+      (controller) => controller.dispose(),
+    );
+
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) => Scaffold(
+        // The Stack is what allows us to retain state across tab
+        // switches by keeping all of our views in the widget tree.
+        body: Stack(
+          fit: StackFit.expand,
+          children: materialNavigationBarItems
+              .map(
+                (barItem) => _buildPageFlow(
+                  context,
+                  materialNavigationBarItems.indexOf(barItem),
+                  barItem,
+                ),
+              )
+              .toList(),
+        ),
+        bottomNavigationBar: BottomNavigationBar(
+          type: BottomNavigationBarType.fixed,
+          backgroundColor: Colors.orangeAccent,
+          selectedItemColor: Colors.white,
+          currentIndex: widget.selectedIndex,
+          items: materialNavigationBarItems
+              .map(
+                (item) => item.bottomNavigationBarItem,
+              )
+              .toList(),
+          onTap: widget.onItemSelected,
+        ),
+      );
+
+  Widget _buildPageFlow(
+    BuildContext context,
+    int tabIndex,
+    _MaterialBottomNavigationTab item,
+  ) {
+    final isCurrentlySelected = tabIndex == widget.selectedIndex;
+
+    // We should build the tab content only if it was already built or
+    // if it is currently selected.
+    _shouldBuildTab[tabIndex] =
+        isCurrentlySelected || _shouldBuildTab[tabIndex];
+
+    final Widget view = FadeTransition(
+      opacity: _animationControllers[tabIndex].drive(
+        CurveTween(curve: Curves.fastOutSlowIn),
+      ),
+      child: KeyedSubtree(
+        key: item.subtreeKey,
+        child: _shouldBuildTab[tabIndex]
+            ? Navigator(
+                // The key enables us to access the Navigator's state inside the
+                // onWillPop callback and for emptying its stack when a tab is
+                // re-selected. That is why a GlobalKey is needed instead of
+                // a simpler ValueKey.
+                key: item.navigatorKey,
+                onGenerateRoute: (settings) => MaterialPageRoute(
+                  settings: settings,
+                  builder: item.initialPageBuilder,
+                ),
+              )
+            : Container(),
+      ),
+    );
+
+    if (tabIndex == widget.selectedIndex) {
+      _animationControllers[tabIndex].forward();
+      return view;
+    } else {
+      _animationControllers[tabIndex].reverse();
+      if (_animationControllers[tabIndex].isAnimating) {
+        return IgnorePointer(child: view);
+      }
+      return Offstage(child: view);
+    }
+  }
+}
+
+/// Extension class of BottomNavigationTab that adds another GlobalKey to it
+/// in order to use it within the KeyedSubtree widget.
+class _MaterialBottomNavigationTab extends BottomNavigationTab {
```

Comment:
I think that you make unnecessary double assertion. I'd leave just a subtreeKey here

Reply:
Kurwa nie wiem czemu pisze to po angielsku, ehh przyzwyczajenia